service/matching/task_reader.go

```diff
@@ -276,6 +276,10 @@ func (tr *taskReader) addTasksToBuffer(
 			// Also increment readLevel for expired tasks otherwise it could result in
 			// looping over the same tasks if all tasks read in the batch are expired
 			tr.backlogMgr.taskAckManager.setReadLevel(t.GetTaskId())
+			// Decreasing the approximateBacklogCounter as this task will no longer be part of the backlog and
```

Comment:
Hmm, I understand what you mean but I see how this doesn't respect the 'present-in-db' semantics. However, this represents a "grey" area since it's a possible count divergence issue which **can** be resolved by this logic.

Consider the case where we have tasks 10, 20, 30 (backlog_counter = 3) where task_ID 20 has expired and the control loop has entered the current function. Now, tasks 10 and 30 will be added to the `outstandingMap` inside of ack_Manager whereas 20 won't be added at all.  

Say Tasks 10 and 30 get completed. Since there were only these two tasks in the map, the ack_level now reads 30 and the counter has decremented by 2, reading 1. When task_gc runs, yes, it will remove task 20 since the ack_level is 30 and it removes all tasks less than 30. However, there will be a count divergence issue which will only be fixed by the resetting logic.

Thus, my question: This case of count divergence seems like something we can "remove".  Do we still want to stick with our semantic, since there could be cases (as described above) where we could be overcounting?

Reply:
interesting example. this solves the example you mentioned, but it has the same problem as decrementing counter after `m.outstandingTasks.Put(taskID, true)` instead of after `m.outstandingTasks.Remove(min)`.
I.e. when we persist ack level if we also persist decrements related to tasks after that ack level, we'll re-read those tasks in case re-loading the queue and we'll decrement for those tasks again, leading to under-count.

I think it's better to have a "looser" overcount and rely on reset, rather than having a tighter count which may be over- or under-counting.